### PR TITLE
[SR] Introduce MapObjectReader

### PR DESCRIPTION
**Pull request by romtsn**
_#skip-changelog_

Add new sentry-android-replay module## :scroll: Description

* Extract an interface `ObjectReader` from `JsonObjecteReader` and introduce a new `MapObjectReader` analogous to what @markushi has done for the hybrid SDKs in https://github.com/getsentry/sentry-java/pull/2814
  * This is necessary to transform object from Maps to typed object for RRWebEvents since we hold a list of generics and lose their type at runtime.
* I also think this probably will be useful in the future if hybrid sdks will pass the data to the native ones in form of dictionaries (probably gonna be the case for Flutter session replay)

There are still some rough edges I believe, but it works for the current replay usecase, and I think it should be fine. I'll add some tests but won't cover everything (will fix stuff along-the-way if I find something).

## Next Steps
Add some tests for the MapObjectReader

### CODE REVIEW

File: sentry/src/main/java/io/sentry/Breadcrumb.java [hunk 1]
```diff
@@ -86,8 +86,7 @@ public static Breadcrumb fromMap(
       switch (entry.getKey()) {
         case JsonKeys.TIMESTAMP:
           if (value instanceof String) {
-            Date deserializedDate =
-                JsonObjectReader.dateOrNull((String) value, options.getLogger());
+            Date deserializedDate = ObjectReader.dateOrNull((String) value, options.getLogger());
```
Comment (markushi) [inline sentry/src/main/java/io/sentry/Breadcrumb.java:89]:
```suggestion
            final Date deserializedDate = ObjectReader.dateOrNull((String) value, options.getLogger());
```

File: sentry/src/main/java/io/sentry/ObjectReader.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import java.io.IOException;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.TimeZone;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface ObjectReader extends Closeable {
+  static @Nullable Date dateOrNull(@Nullable String dateString, final @NotNull ILogger logger) {
```
Comment (markushi) [inline sentry/src/main/java/io/sentry/ObjectReader.java:14]:
```suggestion
  static @Nullable Date dateOrNull(final @Nullable String dateString, final @NotNull ILogger logger) {
```

File: sentry/src/main/java/io/sentry/util/MapObjectWriter.java [hunk 3]
```diff
@@ -151,6 +156,12 @@ public MapObjectWriter value(final @Nullable String value) throws IOException {
     return this;
   }
 
+  @Override
+  public ObjectWriter jsonValue(@Nullable String value) throws IOException {
```
Comment (markushi) [inline sentry/src/main/java/io/sentry/util/MapObjectWriter.java:160]: What's the reason for adding this?

File: sentry/src/main/java/io/sentry/util/MapObjectReader.java [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+import java.util.Date;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.TimeZone;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+@SuppressWarnings("unchecked")
+public final class MapObjectReader implements ObjectReader {
```
Comment (markushi) [inline sentry/src/main/java/io/sentry/util/MapObjectReader.java:20]: there are quite a few local variables to add the `final` keyword to

Comment (markushi) [inline sentry/src/main/java/io/sentry/util/MapObjectReader.java:20]: You could add a simple test where you go full circle: e.g. serialize a context to a map, and the deserialize it back and see if it matches

**Review by markushi — APPROVED** [review]
LGTM, nice!

File: sentry/src/main/java/io/sentry/util/MapObjectReader.java (hunk 4 above)
Comment (romtsn) [reply sentry/src/main/java/io/sentry/util/MapObjectReader.java:21]: yeah I was wrapping it up yday https://github.com/getsentry/sentry-java/pull/3215/commits/1e76fc7499e8a1f4747c67c3aed4d1547c83759e

File: sentry/src/main/java/io/sentry/util/MapObjectWriter.java (hunk 3 above)
Comment (romtsn) [reply sentry/src/main/java/io/sentry/util/MapObjectWriter.java]: sorry this sneaked in from the target branch, I've updated it now to remove from this PR

File: sentry/src/main/java/io/sentry/util/MapObjectReader.java (hunk 4 above)
Comment (romtsn) [reply sentry/src/main/java/io/sentry/util/MapObjectReader.java:21]: added in https://github.com/getsentry/sentry-java/pull/3215/commits/86baf7fc968a8ccc7dc7e8bdce22bd37e60bf8a6

**Pull request merged**